seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.seatunnel.api.configuration;
+
+import static org.apache.seatunnel.api.configuration.util.ConfigUtil.convertToJsonString;
+import static org.apache.seatunnel.api.configuration.util.ConfigUtil.convertValue;
+import static org.apache.seatunnel.api.configuration.util.ConfigUtil.flatteningMap;
+import static org.apache.seatunnel.api.configuration.util.ConfigUtil.treeMap;
+
+import org.apache.seatunnel.shade.com.typesafe.config.Config;
+import org.apache.seatunnel.shade.com.typesafe.config.ConfigRenderOptions;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public class ReadonlyConfig {
+
+    private static final ObjectMapper JACKSON_MAPPER = new ObjectMapper();
+
+    /**
+     * Stores the concrete key/value pairs of this configuration object.
+     */
+    protected final Map<String, Object> confData;
+
+    private ReadonlyConfig(Map<String, Object> confData) {
+        this.confData = confData;
+    }
+
+    public static ReadonlyConfig fromMap(Map<String, Object> map) {
+        return new ReadonlyConfig(treeMap(map));
+    }
+
+    public static ReadonlyConfig fromConfig(Config config) {
+        try {
+            return fromMap(JACKSON_MAPPER.readValue(
+                config.root().render(ConfigRenderOptions.concise()),
+                new TypeReference<Map<String, Object>>() {
+                }));
+        } catch (JsonProcessingException e) {
+            throw new IllegalArgumentException("Json parsing exception.");
+        }
+    }
+
+    public <T> T get(Option<T> option) {
+        return getOptional(option).orElseGet(option::defaultValue);
+    }
+
+    public Map<String, String> toMap() {
+        synchronized (this.confData) {
+            Map<String, Object> flatteningMap = flatteningMap(confData);
+            Map<String, String> result = new HashMap<>(flatteningMap.size());
```

Comment:
`flatteningMap.size()` will convert to array size in map, not map can store data size. If you want set size in hashmap, please use 
```suggestion
            Map<String, String> result = new HashMap<>((float)flatteningMap.size()/0.75 + 1);
```
Or just remove set capacity (recommend).

Reply:
I don't understand, this method doesn't change the size of map keys.

Reply:
Check this: https://cloud.tencent.com/developer/article/1913508

Reply:
You're absolutely right. Thanks